File: backend-python/app/config/init_database.py
import os
import sys
from datetime import datetime, timedelta
import random

# Add parent directory to path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from dotenv import load_dotenv
from app.config.database import Database
from app.models.admin import Admin
from app.models.user import User
from app.models.test_result import TestResult

load_dotenv()

def initialize_database():
    """Initialize database with sample data"""
    try:
        print('\n🔄 Initializing database...\n')
        
        # Initialize database connection
        Database.initialize()
        
        # Create default admin
        print('Creating default admin...')
        try:
            admin_id = Admin.create(
                username=os.getenv('ADMIN_USERNAME', 'admin'),
                email=os.getenv('ADMIN_EMAIL', 'admin@mocktest.com'),
                password=os.getenv('ADMIN_PASSWORD', 'admin123'),
                role='superadmin'
            )
            print(f'✅ Default admin created')
            print(f'   Username: {os.getenv("ADMIN_USERNAME", "admin")}')
            print(f'   Email: {os.getenv("ADMIN_EMAIL", "admin@mocktest.com")}')
        except ValueError as e:
            print(f'ℹ️  Admin already exists')
        
        # Create sample users
        print('\nCreating sample users...')
        sample_users = [
            {
                'name': 'Rahul Kumar',
                'email': 'rahul@example.com',
                'phone': '9876543210',
                'exam_category': 'uttarakhand',
                'tests_completed': 24,
                'total_score': 1872,
                'best_score': 85
            },
            {
                'name': 'Priya Sharma',
                'email': 'priya@example.com',
                'phone': '9876543211',
                'exam_category': 'uttar-pradesh',
                'tests_completed': 18,
                'total_score': 1404,
                'best_score': 82
            },
            {
                'name': 'Amit Patel',
                'email': 'amit@example.com',
                'phone': '9876543212',
                'exam_category': 'ssc',
                'tests_completed': 22,
                'total_score': 1694,
                'best_score': 91
            },
            {
                'name': 'Sneha Verma',
                'email': 'sneha@example.com',
                'phone': '9876543213',
                'exam_category': 'uttarakhand',
                'tests_completed': 16,
                'total_score': 1232,
                'best_score': 89
            },
            {
                'name': 'Vikas Singh',
                'email': 'vikas@example.com',
                'phone': '9876543214',
                'exam_category': 'ssc',
                'tests_completed': 30,
                'total_score': 2850,
                'best_score': 95
            }
        ]
        
        user_count = User.count()
        if user_count == 0:
            user_ids = []
            for user_data in sample_users:
                user_id = User.create(
                    name=user_data['name'],
                    email=user_data['email'],
                    phone=user_data['phone'],
                    exam_category=user_data['exam_category']
                )
                user_ids.append(user_id)
                
                # Update test stats
                collection = User.get_collection()
                collection.update_one(
                    {'_id': user_id},
                    {
                        '$set': {
                            'testsCompleted': user_data['tests_completed'],
                            'totalScore': user_data['total_score'],
                            'bestScore': user_data['best_score'],
                            'averageScore': user_data['total_score'] // user_data['tests_completed']
                        }
                    }
                )
            
            print(f'✅ Created {len(sample_users)} sample users')
        else:
            print(f'ℹ️  {user_count} users already exist')
            # Get existing user IDs for test results
            users = User.find_all(limit=5)
            user_ids = [user['_id'] for user in users]
        
        # Create sample test results
        print('\nCreating sample test results...')
        result_count = TestResult.get_collection().count_documents({})
        
        if result_count == 0 and user_ids:
            results_created = 0
            for user_id in user_ids:
                for i in range(3):
                    answered = 45 + random.randint(0, 5)
                    correct = 30 + random.randint(0, 15)
                    wrong = answered - correct
                    percentage = (correct * 100) // 50
                    
                    TestResult.create(
                        user_id=str(user_id),
                        exam_category='uttarakhand',
                        subject_id='general-studies',
                        subject_name='General Studies',
                        total_questions=50,
                        answered_questions=answered,
                        correct_answers=correct,
                        wrong_answers=wrong,
                        score=percentage,
                        percentage=percentage,
                        time_taken=2400 + random.randint(0, 1200)
                    )
                    results_created += 1
            
            print(f'✅ Created {results_created} sample test results')
        else:
            print(f'ℹ️  {result_count} test results already exist')
        
        print('\n✅ Database initialization complete!\n')
        print('📝 You can now login with:')
        print(f'   Username: {os.getenv("ADMIN_USERNAME", "admin")}')
        print(f'   Password: {os.getenv("ADMIN_PASSWORD", "admin123")}\n')
        
    except Exception as e:
        print(f'❌ Error initializing database: {str(e)}')
        raise e

if __name__ == '__main__':
    initialize_database()

File: backend/src/config/initDatabase.js
require('dotenv').config();
const mongoose = require('mongoose');
const Admin = require('../models/Admin');
const User = require('../models/User');
const TestResult = require('../models/TestResult');

const connectDB = async () => {
  try {
    await mongoose.connect(process.env.MONGODB_URI, {
      useNewUrlParser: true,
      useUnifiedTopology: true,
    });
    console.log('✅ MongoDB Connected');
  } catch (error) {
    console.error('❌ MongoDB connection error:', error);
    process.exit(1);
  }
};

const initializeDatabase = async () => {
  try {
    await connectDB();

    console.log('\n🔄 Initializing database...\n');

    // Create default admin if not exists
    const adminExists = await Admin.findOne({ username: process.env.ADMIN_USERNAME });
    
    if (!adminExists) {
      const admin = await Admin.create({
        username: process.env.ADMIN_USERNAME || 'admin',
        email: process.env.ADMIN_EMAIL || 'admin@mocktest.com',
        password: process.env.ADMIN_PASSWORD || 'admin123',
        role: 'superadmin'
      });
      console.log('✅ Default admin created');
      console.log(`   Username: ${admin.username}`);
      console.log(`   Email: ${admin.email}`);
    } else {
      console.log('ℹ️  Admin already exists');
    }

    // Create sample users
    const userCount = await User.countDocuments();
    
    if (userCount === 0) {
      const sampleUsers = [
        {
          name: 'Rahul Kumar',
          email: 'rahul@example.com',
          phone: '9876543210',
          examCategory: 'uttarakhand',
          testsCompleted: 24,
          totalScore: 1872,
          bestScore: 85
        },
        {
          name: 'Priya Sharma',
          email: 'priya@example.com',
          phone: '9876543211',
          examCategory: 'uttar-pradesh',
          testsCompleted: 18,
          totalScore: 1404,
          bestScore: 82
        },
        {
          name: 'Amit Patel',
          email: 'amit@example.com',
          phone: '9876543212',
          examCategory: 'ssc',
          testsCompleted: 22,
          totalScore: 1694,
          bestScore: 91
        },
        {
          name: 'Sneha Verma',
          email: 'sneha@example.com',
          phone: '9876543213',
          examCategory: 'uttarakhand',
          testsCompleted: 16,
          totalScore: 1232,
          bestScore: 89
        },
        {
          name: 'Vikas Singh',
          email: 'vikas@example.com',
          phone: '9876543214',
          examCategory: 'ssc',
          testsCompleted: 30,
          totalScore: 2850,
          bestScore: 95
        }
      ];

      for (let userData of sampleUsers) {
        const user = await User.create(userData);
        user.calculateAverageScore();
        await user.save();
      }

      console.log(`✅ Created ${sampleUsers.length} sample users`);
    } else {
      console.log(`ℹ️  ${userCount} users already exist`);
    }

    // Create sample test results
    const resultCount = await TestResult.countDocuments();
    
    if (resultCount === 0) {
      const users = await User.find().limit(5);
      
      if (users.length > 0) {
        const sampleResults = [];
        
        users.forEach(user => {
          for (let i = 0; i < 3; i++) {
            sampleResults.push({
              userId: user._id,
              examCategory: user.examCategory,
              subjectId: 'general-studies',
              subjectName: 'General Studies',
              totalQuestions: 50,
              answeredQuestions: 45 + Math.floor(Math.random() * 5),
              correctAnswers: 30 + Math.floor(Math.random() * 15),
              wrongAnswers: 10 + Math.floor(Math.random() * 10),
              score: 60 + Math.floor(Math.random() * 35),
              percentage: 60 + Math.floor(Math.random() * 35),
              timeTaken: 2400 + Math.floor(Math.random() * 1200),
              completedAt: new Date(Date.now() - Math.random() * 7 * 24 * 60 * 60 * 1000)
            });
          }
        });

        await TestResult.insertMany(sampleResults);
        console.log(`✅ Created ${sampleResults.length} sample test results`);
      }
    } else {
      console.log(`ℹ️  ${resultCount} test results already exist`);
    }

    console.log('\n✅ Database initialization complete!\n');
    console.log('📝 You can now login with:');
    console.log(`   Username: ${process.env.ADMIN_USERNAME || 'admin'}`);
    console.log(`   Password: ${process.env.ADMIN_PASSWORD || 'admin123'}\n`);

    process.exit(0);
  } catch (error) {
    console.error('❌ Error initializing database:', error);
    process.exit(1);
  }
};

initializeDatabase();
